### PR TITLE
feat: make usdt/0 great

### LINKED ISSUE #11501: Spike: Can we add Plasma USDT0 under USDT0 related assets
⚠️ **SPIKE - Time-boxed to 2-3 hours max**

⚠️ **Blocked by [Plasma go-live](https://github.com/shapeshift/web/pull/11499)**

## Overview

Currently, Plasma USDT0 (`eip155:9745/erc20:0xb8ce59fc3717ada4c02eadf9682a9e934f625ebb`) is explicitly excluded from related asset grouping in our asset generation script. This was implemented in PR #11369 as a workaround to fix asymmetric grouping issues caused by CoinGecko data corruption.

**The Problem:**

Following Plasma's launch, Plasma USDT0 is now the **canonical primary asset** for the USDT0 family (per CoinMarketCap - Plasma is Tether's stablecoin L1). However, our current fix **excludes the primary from its own group**:

- Plasma USDT0 has `relatedAssetKey: null` (excluded from all grouping)
- Other USDT0 variants are orphaned without their canonical primary
- When Plasma goes live, users will see asymmetric behavior (same bug as #11113):
  - Click Plasma USDT0 → no related assets shown
  - Click other USDT0 variants → Plasma not listed as related

**The Goal:**

Implement proper conflict detection to:
1. Keep USDT and USDT0 as **separate groups** (they are different tokens per CoinGecko)
2. Allow Plasma USDT0 to be the **primary** of the USDT0 group
3. Prevent contract address "stealing" between groups (the root cause)
4. Maintain bidirectional grouping consistency (no asymmetry)

## References and additional details

**Relevant Issues/PRs:**
- Issue #11113 - Original asymmetric grouping problem
- PR #11369 - Current fix that excludes Plasma USDT0
- Issue #11370 - Created to track reverting the exclusion once Plasma is live

**Code Location:**
`scripts/generateAssetData/generateAssetData/generateAssetData.ts`

Plasma USDT0 is currently excluded in **5 locations**:
- Line 142: Filtered from Zerion results
- Line 173: Filtered from CoinGecko results  
- Line 197: Early return in `processRelatedAssetIds()`
- Line 289: Filtered from merged arrays
- Line 347: Cleared in cleanup phase

**Root Cause of Original Bug (#11113):**

CoinGecko has corrupt/cross-contaminated data where USDT (`tether`) and USDT0 (`usdt0`) claim each other's contracts:

**Verified via public APIs (December 2025):**
- **Arbitrum `0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9`**: 
  - Blockchain truth (Arbiscan): USD₮0 token
  - CoinGecko `usdt0`: ✅ Correctly claims it
  - CoinGecko `tether`: ❌ No longer lists (may have been corrected)
  
- **Polygon `0xc2132d05d31c914a87c6611c10748aeb04b58e8f`**:
  - Likely real USDT contract
  - CoinGecko `usdt0`: ❌ Incorrectly claims it
  - Cross-contamination between the two coin IDs

Processing order by market cap causes the issue:
1. USDT (rank ~3) processes first → claims contracts
2. USDT0 (rank ~5000) processes later → **overwrites** `relatedAssetKey` for shared contracts
3. Result: Index has old references but asset metadata is updated → **asymmetry**

**Proposed Solution (from robot analysis in #11113):**

Implement **first-come-first-served conflict detection**:

```typescript
// In processRelatedAssetIds(), before assigning relatedAssetKey:
const cleanedRelatedAssetIds = mergedRelatedAssetIds.filter(candidateAssetId => {
  const existingKey = assetData[candidateAssetId]?.relatedAssetKey
  
  if (!existingKey || existingKey === relatedAssetKey) {
    return true  // No conflict or already in this group
  }
  
  // Conflict: already claimed by different group
  console.warn(
    `[Related Asset Conflict] ${candidateAssetId} already belongs to group ${existingKey}, ` +
    `refusing to add to ${relatedAssetKey}. CoinGecko data quality issue.`
  )
  return false  // Reject - first claim wins
})
```

**Expected Result:**
- USDT (rank ~3) claims shared contracts first
- USDT0 (rank ~5000) tries to claim them → conflict detected → rejected
- Plasma USDT0 becomes primary of USDT0 group with only unique contracts
- Clean separation, no overlap, bidirectional consistency

## Acceptance Criteria

This is a **research spike** - deliverable is feasibility assessment and implementation plan:

1. **Analyze** the conflict detection approach:
   - Where exactly should it be implemented?
   - Does it handle all edge cases?
   - What are the performance implications?

2. **Identify all Plasma exclusion points** and verify they can be safely removed after conflict detection is in place

3. **Test strategy**:
   - How to verify USDT and USDT0 stay separate?
   - How to verify no contract overlap?
   - How to verify bidirectional consistency?
   - How to test with Plasma flag on/off?

4. **Risk assessment**:
   - What could break?
   - Are there other tokens with similar conflicts?
   - What happens to existing user data/state?

5. **Document findings** with:
   - Can it be done safely? (Yes/No/Maybe)
   - Specific implementation steps
   - Estimated effort for full implementation
   - Any gotchas or edge cases discovered

**Success Criteria:**
- Clear go/no-go decision with technical rationale
- If go: Concrete implementation plan with risk mitigation
- If no-go: Alternative recommendations

## Need By Date

Not applicable - research spike

## Screenshots/Mockups

Not applicable

## Estimated effort

2-3 hours (time-boxed - do not exceed)

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

(not again but finally).

This PR:

- ensures proper heuristics to ungroup Coingecko's grouping of USDT/0 
- does a *full* regen from scratch 
- does *another* full regen (not from scratch) on top of it to confirm things will still work following this

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/11501

## Risk

> High Risk PRs Require 2 approvals

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

- There is now a proper USDT0 group (with Plasma chain being the primary asset, Arb, Polygon, Optimism variants also present as well as HyperEVM, Plasma, Monad now added, obviously depending on flag support)
- Plasma USDT0 is now part of the USDT0 group
- USDT group is decoupled from USDT0 (two separate groups)
- https://github.com/shapeshift/web/issues/11113 is *still* fixed, *not* with the wrong current develop way of grouping USDT/0 and excluding Plasma USDT0 from related regen altogether, but by USDT/0 groups being clearly separate *without* any pollution from one ot the other
- Do a general related/assets test, with and without wallet connected

### Engineering

- ^
- worth doing yet another regen to triple-check things on top of the double-check regen 🙏🏽 

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

- ^

## Screenshots (if applicable)

https://jam.dev/c/a5d5b510-bdf7-453b-a162-a12736b2f86f

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved accuracy of related asset detection and grouping with enhanced conflict resolution
  * Fixed potential issues with asset variants being incorrectly mixed in related asset suggestions
  * Added automated data cleanup to remove stale related asset references
  * Strengthened validation to maintain data integrity across asset relationships

* **Chores**
  * Added migration to update internal asset data structures

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 1]
```diff
@@ -91,8 +91,6 @@ const chunkArray = <T>(array: T[], chunkSize: number) => {
   return result
 }
 
-const PLASMA_USDT0_ASSET_ID = 'eip155:9745/erc20:0xb8ce59fc3717ada4c02eadf9682a9e934f625ebb'
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:94]: 👋🏽  all plasma USDT0 exclusion diff

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 2]
```diff
@@ -286,10 +267,43 @@ const processRelatedAssetIds = async (
       ...coingeckoRelatedAssetIds,
       assetId,
     ]),
-  ).filter(id => id !== PLASMA_USDT0_ASSET_ID) // Filter out Plasma USDT0 from final merged array
+  )
+    // Filter to prevent USDT <-> USDT0 cross-contamination
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:271]: Note - all the diff below does *not* mean we'll blindly take USDT/0-named symbols and consider them as being part of the USDT/0 group, we just use this as *additional* filtering for *known* (Zerion/Coingecko data) groups.

**Comment by gomesalexandre** [comment]
Moving to draft as a stack is coming making things even better here but may or may not unify into a single PR.

**Comment by gomesalexandre** [comment]
Superceded by https://github.com/shapeshift/web/issues/11501
